### PR TITLE
Issue #272 Verify the user provide bundle against released one

### PR DESCRIPTION
**Pull request by praveenkumar**
Most of the time user using older version of crc bundle with new
release and this give an unexpected error to the user. We always
need to use the bundle version which released as part of the release.
This check make sure user is going to use the updated one.

Minor refactor around getting bundle info as part of version.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 BUNDLE_VERSION = 4.1.0
-CRC_VERSION = 0.87.0-alpha-$(BUNDLE_VERSION)
+CRC_VERSION = 0.87.0-alpha
```
Comment (gbraad) [inline Makefile]: CRC releases will somehow always be tied to a version, as we need to make sure a bundle can ebe brought up with the associated CRC version. Else we would end up having to test a multitude of version combinations. What is the idea here?

File: cmd/crc/cmd/version.go [hunk 2]
```diff
@@ -37,5 +37,5 @@ var versionCmd = &cobra.Command{
 }
 
 func runPrintVersion(arguments []string) {
-	fmt.Printf("version: %s+%s\n", crcPkg.GetCRCVersion(), crcPkg.GetCommitSha())
+	fmt.Printf("version: %s-%s+%s\n", crcPkg.GetCRCVersion(), crcPkg.GetBundleVersion(), crcPkg.GetCommitSha())
```
Comment (gbraad) [inline cmd/crc/cmd/version.go]: you want to be able to debug the version of the bundle used.... but this indicates another issue too,.. the CRC binary is not independent of the actual bundle it can support.

File: pkg/crc/machine/machine.go [hunk 3]
```diff
@@ -53,14 +54,26 @@ func Start(startConfig StartConfig) (StartResult, error) {
...
 		return *result, err
 	}
 
+	// Verify, user is using latest bundle with released version of crc.
+	if constants.DefaultHostname != crcBundleMetadata.Nodes[0].Hostname {
+		hypervisor := machineConfig.VMDriver
+		if hypervisor == "virtualbox" {
+			hypervisor = "vbox"
+		}
+		errStr := fmt.Sprintf("%s bundle is not supported for this release use update one (crc_%s_%s.tar.xz)", filepath.Base(machineConfig.BundlePath), hypervisor, crc.GetBundleVersion())
```
Comment (gbraad) [inline pkg/crc/machine/machine.go]: confirmed here...

s/update/updated

File: pkg/crc/machine/machine.go [hunk 4]
```diff
@@ -53,14 +54,26 @@ func Start(startConfig StartConfig) (StartResult, error) {
...
+	logging.InfoF("Extracting the %s Bundle tarball ...", filepath.Base(machineConfig.BundlePath))
 	crcBundleMetadata, extractedPath, err := bundle.GetCrcBundleInfo(machineConfig)
 	if err != nil {
 		logging.ErrorF("Error to get bundle Metadata %v", err)
 		result.Error = err.Error()
 		return *result, err
 	}
 
+	// Verify, user is using latest bundle with released version of crc.
+	if constants.DefaultHostname != crcBundleMetadata.Nodes[0].Hostname {
```
Comment (gbraad) [inline pkg/crc/machine/machine.go]: this is not a valid check, as it assume a single situation.

File: pkg/crc/machine/machine.go [hunk 5]
```diff
@@ -53,14 +54,26 @@ func Start(startConfig StartConfig) (StartResult, error) {
...
+		hypervisor := machineConfig.VMDriver
+		if hypervisor == "virtualbox" {
+			hypervisor = "vbox"
```
Comment (gbraad) [inline pkg/crc/machine/machine.go]: why is this necessary?

File: pkg/crc/machine/machine.go (hunk 4 above)
Comment (gbraad) [reply pkg/crc/machine/machine.go]: feels more like a semantic version scheme is needed...

File: pkg/crc/version.go [hunk 6]
```diff
@@ -23,6 +23,9 @@ var (
 
 	// The SHA-1 of the commit this binary is build off
 	commitSha = "sha-unset"
+
+	// Bundle version which used for the release.
+	bundleVersion = "0.0.0-unset"
```
Comment (gbraad) [inline pkg/crc/version.go]: OK, here we set the value of the bundle statically for the binary

File: cmd/crc/cmd/version.go (hunk 2 above)
Comment (gbraad) [reply cmd/crc/cmd/version.go]: So a static version is used... but it is not enforced in the code?

File: pkg/crc/machine/machine.go (hunk 4 above)
Comment (gbraad) [reply pkg/crc/machine/machine.go]: OK, as I understand... you check the cached/running instance node hostname to the one from the bundle... right? If so, make this more obvious by extracting this in a named function,

**Comment by gbraad** [comment]
the intention of what is described in the first message is not clear from the code. I would suggest to extract some of the logic in more clerly named functions; `ValidateVersionInstanceWithImageBundle`, or `ValidateBundleVersions() { return GetInstanceBundleVersion() == GetImageBundleVersion() }` or something similar.

**Comment by gbraad** [comment]
Also not sure why the virtualdriver is overriden here from `virtualbox` to `vbox` ?

File: Makefile (hunk 1 above)
Comment (praveenkumar) [reply Makefile]: @gbraad idea is every release support the released version of bundle only (not older or newer). we still are struggling with the 30 days cert rotation which is not yet resolved for us. Most of the user are trying a older bundle version with latest version of crc and getting issue around the cert expire.

File: cmd/crc/cmd/version.go (hunk 2 above)
Comment (praveenkumar) [reply cmd/crc/cmd/version.go]: @gbraad I didn't get that comment, we had it part of Makefile till now and this is just I made it part of codebase.

File: pkg/crc/machine/machine.go (hunk 5 above)
Comment (praveenkumar) [reply pkg/crc/machine/machine.go]: @gbraad because our bundle contain `vbox` instead `virtualbox` in the tarball, we can either change it there, make part of snc script or here.

Comment (gbraad) [reply pkg/crc/machine/machine.go]: this is not the place to change/override it... this feels like a workaround value.

@code-ready/crc-devel WDYT?

File: cmd/crc/cmd/version.go (hunk 2 above)
Comment (gbraad) [reply cmd/crc/cmd/version.go]: See https://github.com/code-ready/crc/pull/273#issuecomment-507158507

File: pkg/crc/machine/machine.go (hunk 5 above)
Comment (praveenkumar) [reply pkg/crc/machine/machine.go]: @gbraad yes this is workaround, I will create an issue on snc side so we can generated the tarball with hypervisor next time, Will put Note here meanwhile.

Comment (praveenkumar) [reply pkg/crc/machine/machine.go]: @gbraad https://github.com/code-ready/snc/issues/56 here

Comment (gbraad) [reply pkg/crc/machine/machine.go]: If this is approved and merged, will we eventually 'remove' this with the next release?

Comment (praveenkumar) [reply pkg/crc/machine/machine.go]: @gbraad yes we will remove that part from next release.

File: pkg/crc/validation/validation.go [hunk 7]
```diff
@@ -39,5 +43,11 @@ func ValidateBundle(bundle string) error {
 	if _, err := os.Stat(bundle); os.IsNotExist(err) {
 		return errors.NewF("Expected file %s does not exist", bundle)
 	}
+	// Check if the version of the bundle provided by user is same as what is released with crc.
+	releaseBundleVersion := crc.GetBundleVersion()
+	userProvidedBundleVersion := filepath.Base(bundle)
+	if !strings.Contains(userProvidedBundleVersion, fmt.Sprintf("%s.tar.xz", releaseBundleVersion)) {
```
Comment (gbraad) [inline pkg/crc/validation/validation.go]: slightly harder to read, but you are constructing the bundle path to compare.
